pyproject.toml

```diff
@@ -83,8 +65,41 @@ git-email = "52496925+botpub@users.noreply.github.com"
 changelog-file = "docs/changelog.rst"
 changelog-header = "###############"
 version-header = "="
-version-strings = ["setup.py"]
-build-system = "setuptools"
+
+[tool.pdm]
+
+[tool.pdm.scripts]
+docbuild = "invoke docbuild"
+docserve = "invoke docserve"
+lint = "invoke lint"
+test = "invoke tests"
+
+[tool.pdm.dev-dependencies]
+dev = [
+    "BeautifulSoup4<5.0,>=4.9",
+    "jinja2~=3.1.2",
+    "lxml<5.0,>=4.3",
+    "markdown~=3.4.3",
+    "typogrify<3.0,>=2.0",
+    "sphinx<6.0,>=5.1",
+    "furo==2023.03.27",
+    "livereload<3.0,>=2.6",
+    "psutil<6.0,>=5.7",
+    "pygments~=2.15",
+    "pytest<8.0,>=7.1",
+    "pytest-cov<5.0,>=4.0",
+    "pytest-sugar<1.0.0,>=0.9.5",
+    "pytest-xdist<3.0,>=2.0",
+    "tox<4.0,>=3.13",
+    "flake8<4.0,>=3.8",
+    "flake8-import-order<1.0.0,>=0.18.1",
+    "invoke<3.0,>=2.0",
+    "isort<6.0,>=5.2",
+    "black<20.0,>=19.10b0",
+    "ruff>=0.1.3,<1.0.0",
+    "tomli;python_version<'3.11'",
+]
```

Comment:
1. Should we replace the long `dev` group with a more-specific set of `docs`, `lint`, and `test` groups? ([example](https://github.com/pelican-plugins/featured-image/blob/38dbae18cf684657670f0b375dd1a8a68a5323a5/pyproject.toml#L39-L50))
2. Except for Jinja2, Markdown, Pygments, and Furo, I think we can remove the upper bounds. More often than not, upper bounds cause more problems than they solve. (See example of the [Black project dropping upper bounds on dependencies](https://github.com/psf/black/pull/2718).)
3. Given that we are going to use Ruff, I don't think we need `flake8`, `flake8-import-order`, or `isort` any longer — Ruff contains all the functionality of those tools.

Reply:
yes I think so too and thought about it. To keep this PR a bit smaller I propose to do it in an additional PR

Reply:
I agree that deferring the separation into development dependency groups makes sense, but I think my comments in (2) and (3) are probably best resolved in the context of this PR, don't you think?